### ISSUE TITLE
[11.0] Regresion: Download button for binary fields doesn't work in tree view

### ISSUE THREAD

**Issue by yelizariev** [odoo/odoo#21996]
**Impacted versions:**

* works in 10.0 
* doesn't work in 11.0

**Steps to reproduce:**

* add field ``datas`` to tree view of ``ir.attachment`` model

**Current behavior:**

Shows base64 representation of the file 

**Expected behavior:**

Show Download button

Similar question at odoo forum: https://www.odoo.com/forum/help-1/question/problem-with-tree-view-and-file-field-125752

**Screenshots:**

![odoo-11](https://user-images.githubusercontent.com/186131/34530317-970cbbbe-f0d0-11e7-99aa-dc6281cb0ad8.png)
![odoo-10](https://user-images.githubusercontent.com/186131/34530318-973567a8-f0d0-11e7-8425-0fc7dd385aef.png)

**Comment by Yenthe666** [comment]
@mart-e this is framework related. Can you check please?

**Comment by mart-e** [comment]
@ged-odoo looks like new views-related ;-)

**Comment by ged-odoo** [comment]
@alexkuhn could you have a look when you have some time?

**Comment by alexkuhn** [comment]
Hello, this commit fixes the issue: https://github.com/odoo-dev/odoo/commit/6247d44721908043adbef149bde1c593966a6dcb

By default, a binary field in a list view displays its estimated size (e.g. 2.92 Bytes).
If you would like this field to display a download link, you should use the widget `download_link`.

For instance, with attachments, add the binary field as follow:

`<field name="datas" widget="download_link" options="{'filename': 'datas_fname'}"/>`

The option `filename` sets which field has the filename as its value.
If omitted, the downloaded file will be named "Binary File".

**Comment by yelizariev** [comment]
Thank you @alexkuhn, but I still have few concerns:

* ALL Binary data are downloaded when tree view is opened. It decreases performance on server side as well as client side. There is special option to download  only size of the files from the server, check it please: [bin_size](https://github.com/odoo/odoo/search?l=JavaScript&q=bin_size&type=&utf8=%E2%9C%93)

![s1](https://user-images.githubusercontent.com/186131/34684926-8a4e3c12-f4c8-11e7-9d20-81abb252fce4.png)


* Can we have filename in the link instead of just "Download" label? That would be great. Probably same icon as in form view is also good idea
![s](https://user-images.githubusercontent.com/186131/34683812-00e77c02-f4c5-11e7-9006-df0df69ff296.png)

**Comment by Yenthe666** [comment]
Both suggestions sound very good to me honestly.

**Comment by alexkuhn** [comment]
## 1. Binary field representation in list view

It was already possible to represent a binary field on a list view in a similar way to the form view:

![binary-list-view](https://user-images.githubusercontent.com/6569390/34882091-bb72ec9e-f7b5-11e7-8083-f58697549945.png)

![binary-list-view-arch](https://user-images.githubusercontent.com/6569390/34882097-c1a620cc-f7b5-11e7-8c9c-05a96334dffd.png)

If you don't state any widget, it represents the binary fields by their size in bytes.
The attribute `filename` states which field contains the filename as its value.

So I have removed the "download_link" widget with this commit: 7813a8e714f268983b0be88827ae1c8ad887e6c7

## 2. Performance

You could pass `{ bin_size: true }` in the context to the server, so that it represents the value of binary fields by their size in bytes. As a consequence, the web client does not download the data when the context has set this key.

However, it was a bit annoying that you had to manually set this key in the context of any action that leads to a view with some binary fields.

So as an improvement, whenever you load a view containing a binary field, it sets `{ bin_size: true }` in the context: 41fe7f9d1113d026b226b136b4477aaaf9f1f638

**Comment by alexkuhn** [comment]
I have reverted the changes about bin_size set to true for binary fields, as it was the source of some annoying issues (see #22222).

Here's the commit for the revision: 41fe7f9d1113d026b226b136b4477aaaf9f1f638

To make it short, there were two problems:

1. the server also gets a bin size when it reads a binary field with `bin_size` set in the context (see https://github.com/odoo/odoo/issues/22222#issue-288309373).
2. Some views rely on downloading the binary data when accessing a record (see https://github.com/odoo/odoo/issues/22222#issuecomment-357537695).

The 1st problem might be a bug in the Python code, but it can be hard to fix it as it might be a design flaw (see https://github.com/odoo/odoo/pull/22225#issuecomment-357476107).

As for the 2nd problem, there are too many views that rely on that, and that is without considering custom modules from the community.

**Comment by yelizariev** [comment]
Ok, so bin_size has to be passed manually

**Comment by yelizariev** [comment]
@alexkuhn is it possible to activate bin_size by default for tree views only? As I understand those issues are related for form view?

**Comment by alexkuhn** [comment]
I will investigate to see if it is feasible, without having the above issues.

**Comment by Elkasitu** [comment]
@yelizariev @alexkuhn if it's for performance reasons only, I'd suggest doing it in master and not in a stable branch :^)

**Comment by yelizariev** [comment]
@Elkasitu  It's a kind of performance issue, that makes feature unusable. So, it's a bug, rather than just improvement or something

**Comment by Elkasitu** [comment]
@yelizariev But there is an alternative which is passing `bin_size: True` in the context; there is a chance that enabling bin_size: True by default for tree views might break something else and then we'd have to revert the fix for the third time, which is something nobody wants.

Thankfully Odoo is open source so if you really want to gamble stability for performance, you're free to do it on your own instance.

TL;DR: there are more people who complained about 41fe7f9d1 breaking their shit than people complaining about this specific issue

**Comment by Yenthe666** [comment]
@Elkasitu could we atleast add this information somewhere in the V11 docs?

**Comment by Elkasitu** [comment]
@Yenthe666 tbd with @mart-e perhaps

**Comment by yelizariev** [comment]
I understand, that stability is more important than performance. If there is technical reasons that don't allow to get stable fix of performance issue, then some note or warning has to be added to docs about using binary fields in tree view as @Yenthe666 said
